### PR TITLE
Improved typings for Subscriptions context and authorization

### PR DESCRIPTION
**Pull request by darrellwarde**
# Description

This PR extends the improvements to context typings to also include Subscriptions. This includes removing some duplicity, and standardizing how JWT tokens are decoded and verified.

## Complexity

Complexity: Medium

### CODE REVIEW

File: packages/graphql/src/classes/Neo4jGraphQL.ts [hunk 1]
```diff
@@ -287,11 +288,18 @@ class Neo4jGraphQL {
...
+            "Mutation.*": [wrapQueryAndMutation(wrapResolverArgs)],
         };
 
+        if (this.features.subscriptions) {
+            resolversComposition["Subscription.*"] = wrapSubscription({
+                subscriptionsEngine: this.features.subscriptions,
+                schemaModel: this.schemaModel,
+                authorization: this.authorization,
+            });
+        }
```
Comment (darrellwarde) [inline packages/graphql/src/classes/Neo4jGraphQL.ts:301]: @angrykoala, with this conditional, we can remove the check we discussed from the wrapper, which I think improves the comprehensibility of the code. 🙂

**Comment by neo4j-team-graphql** [comment]
Made some changes to the Neo4j GraphQL Toolbox? Or made some changes to the core library perhaps?

Check out your changes at https://neo4j-graphql-toolbox-3810.surge.sh

Use for instance a local dbms hosted on [Neo4j Desktop](https://neo4j.com/download/) or a cloud-hosted [Aura Free](https://neo4j.com/cloud/aura-free/) dbms.

**Comment by darrellwarde** [comment]
@a-alle, this is pretty much the continuation of the last authorization fix, which gets rid of the duplicity temporarily introduced. 🙂

**Comment by neo4j-team-graphql** [comment]
# Performance Report

No Performance Changes


<details>
    <summary>Show Full Table</summary>
    
| name | dbHits | old dbHits | time (ms) | old time (ms) | maxRows |
| ---- | ------ | ---------- | --------- | ------------- | ------- |
|  aggregations.TopLevelAggregate | 3403 | 3403 | 43 | 45 | 1134 |
|  aggregations.NestedAggregation | 16553 | 16553 | 99 | 92 | 2174 |
|  aggregations.AggregationWithWhere | 11979 | 11979 | 62 | 58 | 2174 |
|  aggregations.AggregationWhereWithinNestedRelationships | 22116987 | 22116987 | 4039 | 3669 | 2008534 |
|  aggregations.AggregationWhereWithinNestedConnections | 22116987 | 22116987 | 3380 | 3277 | 2008534 |
|  aggregations.NestedCountFromMovieToActors | 9734 | 9734 | 45 | 51 | 2174 |
|  aggregations.NestedCountFromActorsToMovie | 9937 | 9937 | 53 | 62 | 2174 |
|  aggregations.DeeplyNestedCount | 13070329 | 13070331 | 6080 | 5778 | 2008534 |
|  batch-create.BatchCreate | 4200 | 4200 | 139 | 192 | 600 |
|  batch-create.BatchCreateSmall | 77 | 77 | 69 | 87 | 11 |
|  connect.createAndConnect | 14424 | 14424 | 288 | 349 | 3003 |
|  connections.Connection | 14082 | 14082 | 76 | 80 | 2174 |
|  connections.NestedConnection | 45472 | 45468 | 134 | 197 | 4516 |
|  connections.ConnectionWithSort | 3284 | 3284 | 79 | 96 | 1040 |
|  connections.ConnectionWithSortAndCypher | 16325 | 16325 | 111 | 130 | 2174 |
|  create.SimpleMutation | 7 | 7 | 70 | 129 | 1 |
|  cypher-directive.TopLevelMutationDirective | 1135 | 1135 | 38 | 43 | 1134 |
|  delete.SimpleDelete | 19401 | 19401 | 293 | 302 | 1040 |
|  delete.NestedDeleteInUpdate | 19198 | 19198 | 322 | 246 | 1183 |
|  2925.SingleRelationshipFilter | 6468 | 6468 | 90 | 57 | 1040 |
|  2925.NestedSingleRelationshipFilter | 22900 | 22900 | 106 | 161 | 2174 |
|  2925.SingleRelationshipRequiredFilter | 5201 | 5201 | 44 | 47 | 1040 |
|  2925.NestedSingleRelationshipRequiredFilter | 9361 | 9361 | 62 | 73 | 1040 |
|  query.SimpleQuery | 3121 | 3121 | 24 | 27 | 1040 |
|  query.SimpleQueryWithRelationship | 16162 | 16162 | 53 | 54 | 2174 |
|  query.QueryWhere | 9706 | 9695 | 44 | 53 | 2165 |
|  query.SimpleQueryWithNestedWhere | 9884 | 9873 | 73 | 89 | 2165 |
|  query.Nested | 10096041 | 10096041 | 11625 | 11200 | 2008534 |
|  query.NestedWithFilter | 10074401 | 10074401 | 11161 | 11211 | 2004000 |
|  query.OrFilterOnRelationships | 42063 | 43146 | 217 | 271 | 1519 |
|  query.OrFilterOnRelationshipsAndNested | 35562 | 37265 | 233 | 354 | 1519 |
|  query.QueryWithNestedIn | 13785 | 13330 | 70 | 95 | 1343 |
|  query.NestedConnectionWhere | 9834 | 9834 | 66 | 78 | 2174 |
|  query.DeeplyNestedConnectionWhere | 10084 | 9896 | 86 | 119 | 2174 |
|  query.DeeplyNestedWithRelationshipFilters | 19308 | 19116 | 168 | 183 | 1652 |
|  query.NestedWithRelationshipSingleFilters | 3881 | 3881 | 236 | 244 | 1134 |
|  query.Fulltext | 80 | 80 | 81 | 40 | 16 |
|  query.FulltextWithNestedQuery | 587 | 587 | 102 | 54 | 84 |
|  sorting.SortMultipleTypes | 3515 | 3515 | 86 | 93 | 1040 |
|  sorting.SortMultipleTypesWithCypherWithCypher | 14527 | 14491 | 108 | 114 | 2174 |
|  sorting.SortOnNestedFields | 14082 | 14082 | 56 | 65 | 2174 |
|  sorting.SortDeeplyNestedFields | 43198 | 43198 | 107 | 126 | 4516 |
|  sorting.SortWithTopLevelCypher | 3396 | 3396 | 72 | 85 | 1040 |
|  unions.SimpleUnionQuery | 321 | 321 | 74 | 88 | 35 |
|  unions.SimpleUnionQueryWithMissingFields | 293 | 293 | 70 | 107 | 35 |
|  unions.NestedUnion | 410637 | 410637 | 312 | 302 | 33033 |
|  unions.NestedUnionWithMissingFields | 384611 | 384611 | 284 | 332 | 33033 |
|  update.NestedUpdate | 16143 | 16143 | 112 | 134 | 2002 |

</details>
        
Old Schema Generation: 50.231s
Schema Generation: 53.563s
Old Subgraph Schema Generation: 1:00.926 (m:ss.mmm)
Subgraph Schema Generation: 1:04.774 (m:ss.mmm)

File: packages/graphql/src/types/neo4j-graphql-subscriptions-context.ts [hunk 2]
```diff
@@ -0,0 +1,26 @@
...
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+import type { Neo4jGraphQLCommonContext } from "./neo4j-graphql-common-context";
+
+export interface Neo4jGraphQLSubscriptionsConnectionParams extends Neo4jGraphQLCommonContext {}
```
Comment (angrykoala) [inline packages/graphql/src/types/neo4j-graphql-subscriptions-context.ts:22]: Isn't it a bit redundant to call the superclass Neo4jGraphQL**Common**Context?

File: packages/graphql/src/schema/resolvers/composition/utils/get-authorization-context.ts [hunk 3]
```diff
@@ -0,0 +1,63 @@
...
+import type { Neo4jGraphQLContext } from "../../../../types/neo4j-graphql-context";
+import type { Neo4jGraphQLSubscriptionsConnectionParams } from "../../../../types/neo4j-graphql-subscriptions-context";
+
+export async function getAuthorizationContext(
+    context: Neo4jGraphQLContext | Neo4jGraphQLSubscriptionsConnectionParams,
+    authorization?: Neo4jGraphQLAuthorization,
+    jwtClaimsMap?: Map<string, string>
+): Promise<AuthorizationContext> {
+    if (!context.jwt) {
+        if (authorization) {
```
Comment (angrykoala) [inline packages/graphql/src/schema/resolvers/composition/utils/get-authorization-context.ts:32]: Any reason for not having a single if?
```suggestion
    if (!context.jwt && authorization) {
```

File: packages/graphql/src/schema/resolvers/composition/utils/get-authorization-context.ts [hunk 4]
```diff
@@ -0,0 +1,63 @@
...
+                };
+            } catch (e) {
+                const isAuthenticated = false;
+                return {
+                    isAuthenticated,
+                    jwtParam: new Cypher.NamedParam("jwt", {}),
+                    isAuthenticatedParam: new Cypher.NamedParam("isAuthenticated", isAuthenticated),
+                };
+            }
+        }
```
Comment (angrykoala) [inline packages/graphql/src/schema/resolvers/composition/utils/get-authorization-context.ts:51]: I see here that we are using errors for flow control. Why is `.decode` throwing an error if we are going to catch it anyway in every case? Could this be a bit cleaner if `decode` returned undefined or something?

**Review by angrykoala — APPROVED** [review]
Nice, just a couple of minor cometns to address before merging

File: packages/graphql/src/types/neo4j-graphql-subscriptions-context.ts (hunk 2 above)
Comment (darrellwarde) [reply packages/graphql/src/types/neo4j-graphql-subscriptions-context.ts:22]: Well `Neo4jGraphQLContext` is the class name we expose to users, so we can't use that, do you want me to call it `INeo4jGraphQLContext`? 😜

File: packages/graphql/src/schema/resolvers/composition/utils/get-authorization-context.ts (hunk 4 above)
Comment (darrellwarde) [reply packages/graphql/src/schema/resolvers/composition/utils/get-authorization-context.ts]: Good spot, addressed this now!

File: packages/graphql/src/types/neo4j-graphql-subscriptions-context.ts (hunk 2 above)
Comment (angrykoala) [reply packages/graphql/src/types/neo4j-graphql-subscriptions-context.ts:22]: Good point. What about `Neo4jGraphQLContextInterface`?

I know it is not the bestest name ever, but I think still beats slightly `Common`

**Pull request merged**